pkg/evs/controllerserver.go

```diff
@@ -264,20 +267,189 @@ func (cs *ControllerServer) ListVolumes(_ context.Context, req *csi.ListVolumesR
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-func (cs *ControllerServer) CreateSnapshot(_ context.Context, req *csi.CreateSnapshotRequest) (
-	*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (cs *ControllerServer) CreateSnapshot(_ context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	klog.V(4).Infof("CreateSnapshot called with request %v", *req)
+	credentials := cs.Driver.cloudCredentials
+	name := req.GetName()
+	volumeId := req.GetSourceVolumeId()
+	if err := checkCreateSnapshotParamEmpty(name, volumeId); err != nil {
+		return nil, err
+	}
+	if response, err := checkRepeatSnapshotName(credentials, name, volumeId); response != nil || err != nil {
+		return response, err
+	}
+	if err := checkVolumeIsExist(credentials, volumeId); err != nil {
+		return nil, err
+	}
+	return doSnapshotCreate(credentials, name, volumeId)
 }
 
-func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (
-	*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func checkCreateSnapshotParamEmpty(name string, volumeId string) error {
+	if volumeId == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot SourceVolumeId cannot be empty")
+	}
+	if name == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot Name cannot be empty")
+	}
+	return nil
 }
 
-func (cs *ControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (
-	*csi.ListSnapshotsResponse,
-	error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func checkRepeatSnapshotName(credentials *config.CloudCredentials, name string, volumeId string) (*csi.CreateSnapshotResponse, error) {
+	listOpts := snapshots.ListOpts{
+		Name: name,
+	}
+	listSnapshots, err := services.ListSnapshots(credentials, listOpts)
+	if err != nil {
+		klog.Errorf("Failed to query snapshot list: %v", err)
+		return nil, status.Error(codes.Internal, "Failed to get snapshots")
```

Comment:
`Failed to query a snapshots list` is better

Reply:
done